src/common/viz/widgets/BrowserAssetWidget.js

```diff
@@ -0,0 +1,157 @@
+            }
+            this.file = null;
+
+            this.logger = Logger.create('deepforge:viz:widgets:BrowserAssetWidget',
+                WebGMEGlobal.gmeConfig.client.log);
+
+            this.parentEl = ASSET_WIDGET_BASE.clone();
+            this.el.append(this.parentEl);
+
+            this.assetLink = ASSET_LINK.clone();
+            this.parentEl.append(this.assetLink);
+
+            this.fileDropTarget = this.parentEl;
+
+            this.btnAttach = BTN_ATTACH.clone();
+            this.parentEl.append(this.btnAttach);
+
+            this.fileUploadInput = INPUT_FILE_UPLOAD.clone();
+
+            this.attachFileDropHandlers();
+            this.updateDisplay();
+        }
+
+        attachFileDropHandlers() {
+            this.btnAttach.on('click', e => {
+                e.stopPropagation();
+                e.preventDefault();
+
+                this.fileUploadInput.click();
+            });
+
+            this.fileUploadInput.on('change', e => {
+                e.stopPropagation();
+                e.preventDefault();
+                this.fileSelectHandler(e.originalEvent);
+            });
+
+            this.fileDropTarget.on('dragover',  e => {
+                e.stopPropagation();
+                e.preventDefault();
+            });
+
+            this.fileDropTarget.on('dragenter', e => {
+                e.stopPropagation();
+                e.preventDefault();
+                this.fileDropTarget.addClass('hover');
+            });
+
+            this.fileDropTarget.on('dragleave', e => {
+                e.stopPropagation();
+                e.preventDefault();
+                this.fileDropTarget.removeClass('hover');
+            });
+
+            this.fileDropTarget.on('drop', e => {
+                e.stopPropagation();
+                e.preventDefault();
+                this.fileDropTarget.removeClass('hover');
+                this.fileSelectHandler(e.originalEvent);
+            });
+        }
+
+        fileSelectHandler(event){
+            const files = event.target.files || event.dataTransfer.files;
+            if(files){
+                this.setFile(files[0]);
+            }
+        }
+
+        setFile(file) {
+            this.file = file;
+            this.assetLink.text(this.file.name);
+            this.assetLink.attr('href', URL.createObjectURL(this.file));
+        }
+
+        detachFileHandlers(){
+            this.fileUploadInput.off('change');
+
+            this.fileDropTarget.off('dragover');
+            this.fileDropTarget.off('dragenter');
+            this.fileDropTarget.off('dragleave');
+            this.fileDropTarget.off('drop');
+
+            this.btnAttach.off('click');
+        }
+
+        updateDisplay() {
+            if(this.file){
+                this.setFile(this.file);
+            }
+            super.updateDisplay();
+        }
+
+        async beforeSubmit(basedir, config) {
+            const storageConfig = findByKey(config, 'storage');
```

Comment:
Ideally, I would actually include the storage config settings in this widget as it shouldn't reach outside of itself to look up peer config values.